### PR TITLE
fix(server): Add additional metrics

### PR DESCRIPTION
**Pull request by dranikpg**
1.  Clean up metrics collection (no functional changes)
    - Group together tx stats in Metrics struct
    - Remove ADD_HEADER
    - Small polishment
2. Add dispatch queue stats
    - entries
    - bytes
    - peak_bytes
3. Add replication buffer stats
    - total streamer buffer capacities
    - full sync buffers & channel bytes
4. Add a per-thread limit to the dispatch queue size

### CODE REVIEW

File: src/facade/dragonfly_connection.cc [hunk 1]
```diff
@@ -923,7 +936,7 @@ void Connection::DispatchFiber(util::FiberSocketBase* peer) {
       recycle(move(msg));
     }
 
-    evc_bp_.notify();
+    queue_backpressure.ec.notify();
```
Comment (dranikpg) [inline src/facade/dragonfly_connection.cc:939]: I'm worried about the overhead of accessing the `thread_local` here in the loop (however pipeline_req_pool_ is thread-local too....)

File: src/facade/facade.cc [hunk 2]
```diff
@@ -27,17 +27,19 @@ constexpr size_t kSizeConnStats = sizeof(ConnectionStats);
...
 ConnectionStats& ConnectionStats::operator+=(const ConnectionStats& o) {
   // To break this code deliberately if we add/remove a field to this struct.
-  static_assert(kSizeConnStats == 136u);
+  static_assert(kSizeConnStats == 152u);
 
   ADD(read_buf_capacity);
-  ADD(pipeline_cache_capacity);
+  ADD(dispatch_queue_entries);
+  ADD(dispatch_queue_bytes);
+  ADD(dispatch_queue_peak_bytes);
```
Comment (romange) [inline src/facade/facade.cc:35]: interesting. global peak bytes are not the sum of peak bytes per connection or thread. I wonder if we can do this over estimation.

File: src/server/main_service.cc [hunk 3]
```diff
@@ -1026,6 +1024,8 @@ bool Service::InvokeCmd(const CommandId* cid, CmdArgList tail_args, ConnectionCo
     DispatchMonitor(cntx, cid, tail_args);
   }
 
+  etl.RecordCmd();
```
Comment (romange) [inline src/server/main_service.cc:1027]: what's the impact and reasons of this change?

File: src/server/dflycmd.h [hunk 4]
```diff
@@ -136,6 +139,8 @@ class DflyCmd {
 
   std::vector<ReplicaRoleInfo> GetReplicasRoleInfo();
 
+  void GetReplicationMemoryStats(ReplicationMemoryStats* out);
```
Comment (romange) [inline src/server/dflycmd.h:142]: const?

File: src/facade/dragonfly_connection.cc [hunk 5]
```diff
@@ -1074,8 +1081,9 @@ void Connection::SendAsync(MessageHandle msg) {
 }
 
 void Connection::EnsureAsyncMemoryBudget() {
-  evc_bp_.await(
-      [this] { return dispatch_q_bytes_.load(memory_order_relaxed) <= kMaxDispatchQMemory; });
+  queue_backpressure_.ec.await([this] {
```
Comment (romange) [inline src/facade/dragonfly_connection.cc:1084]: introduce this function inside the struct itself

File: src/server/main_service.cc (hunk 3 above)
Comment (dranikpg) [reply src/server/main_service.cc:1027]: This just computes the QPS stats. Not all commands are run via DispatchCommand, but I can move it back

File: src/facade/facade.cc (hunk 2 above)
Comment (dranikpg) [reply src/facade/facade.cc:35]: True, the sum of peaks can be much higher even in modest usage scenarios. But then I have to find some place to store it... Will find one

File: src/server/main_service.cc (hunk 3 above)
Comment (dranikpg) [reply src/server/main_service.cc]: Done

File: src/facade/facade.cc (hunk 2 above)
Comment (dranikpg) [reply src/facade/facade.cc]: Fixed

File: src/server/server_family.cc [hunk 6]
```diff
@@ -1202,66 +1207,71 @@ void ServerFamily::Save(CmdArgList args, ConnectionContext* cntx) {
...
 
-  result.is_master = false;
-  if (ServerState::tlocal() && ServerState::tlocal()->is_master) {
-    result.is_master = true;
+  result.is_master = ServerState::tlocal() && ServerState::tlocal()->is_master;
+  if (result.is_master)
     result.replication_metrics = dfly_cmd_->GetReplicasRoleInfo();
-  }
+
+  // Update peak stats
```
Comment (romange) [inline src/server/server_family.cc:1269]: Looks good but please add a comment saying that we rely here on GetMetrics being periodically called.

Comment (dranikpg) [reply src/server/server_family.cc:1269]: Hm actually that's not what we want, do we?

Comment (dranikpg) [reply src/server/server_family.cc:1269]: Maybe we can make them atomic and update them from heartbeat? But then we have to partially assemble the stats 🤯

**Review by romange — APPROVED** [review]

**Pull request merged**